examples/stock/company.v.py

```diff
@@ -27,23 +32,61 @@ def buy_stock():
     buy_order = msg.value / self.price # rounds down
 
     # There are enough shares to buy
-    assert self.holdings[self.company] >= buy_order
+    assert self.stock_available() >= buy_order
     
     # Take the shares off the market and give to stockholder
     self.holdings[self.company] -= buy_order
     self.holdings[msg.sender] += buy_order
 
+# So someone can find out how much they have
+@constant
+def get_holding(_stockholder: address) -> currency_value:
+    return self.holdings[_stockholder]
+
+# The amount the company has on hand in cash
+@constant
+def cash() -> wei_value:
+    return self.balance
+
 # Give stock back to company and get my money back!
 def sell_stock(sell_order: currency_value):
     # Can only sell as much stock as you own
-    assert self.holdings[msg.sender] >= sell_order
+    assert self.get_holding(msg.sender) >= sell_order
+    # Company can pay you
+    assert self.cash() >= (sell_order * self.price)
 
     # Sell the stock, send the proceeds to the user
     # and put the stock back on the market
     self.holdings[msg.sender] -= sell_order
     self.holdings[self.company] += sell_order
     send(msg.sender, sell_order * self.price)
 
+# Transfer stock from one stockholder to another
+# (Assumes the receiver is given some compensation, but not enforced)
+def transfer_stock(receiver: address, transfer_order: currency_value):
+    # Can only trade as much stock as you own
+    assert self.get_holding(msg.sender) >= transfer_order
+    
+    # Debit sender's stock and add to receiver's address
+    self.holdings[msg.sender] -= transfer_order
+    self.holdings[receiver] += transfer_order
+
+# Allows the company to pay someone for services rendered
+def pay_bill(vendor: address, amount: wei_value):
+    # Only the company can pay people
+    assert msg.sender == self.company
+    # And only if there's enough to pay them with
+    assert self.cash() >= amount
+
+    # Pay the bill!
+    send(vendor, amount)
+
+# The amount a company has raised in the stock offering
 @constant
-def valuation() -> wei_value:
+def worth() -> wei_value:
     return (self.total_shares - self.holdings[self.company]) * self.price
+
```

Comment:
@fubuloubu I'd call `worth` something else or change how it's calculated as when I think of a companies worth I think of all the stock including the shares that the company owns.  What do you think?

Reply:
I was thinking about the valuation as the amount owed to every stockholder if all of a sudden they all wanted to sell to the company. I think this whole example is a bit different then how stocks traditionally work in that you don't usually buy and sell directly to the company, this idea was more representative of what could be possible if you changed the paradigm a bit.

That being said, if you have a better name, I'd consider it. It's not really a problem to change it.

Reply:
It seems like the company is beholden to the shareholders for `worth()`, what do you think about calling it `debt()`?

Reply:
I had a debt function below that represent the value held in cash versus the amount held by stock. I think I could have different names if it would make it clearer

Reply:
That'd be great! Thanks